go/consensus/tendermint/tendermint.go

```diff
@@ -1280,6 +1289,49 @@ func (t *tendermintService) worker() {
 	}
 }
 
+// metrics updates oasis_consensus metrics by checking last accepted block info.
+func (t *tendermintService) metrics() {
+	sub, err := t.Subscribe("tendermint/metrics", tmtypes.EventQueryNewBlock)
+	if err != nil {
+		t.Logger.Error("worker: failed to subscribe to new block events",
+			"err", err,
+		)
+		return
+	}
+	defer t.Unsubscribe("tendermint/metrics", tmtypes.EventQueryNewBlock) // nolint:errcheck
+
+	// Tendermint uses specific public key encoding.
+	pubKey := t.consensusSigner.Public()
+	myAddr := []byte(crypto.PublicKeyToTendermint(&pubKey).Address())
+	for {
+		var blk *tmtypes.Block
+		select {
+		case <-t.node.Quit():
+			return
+		case <-sub.Cancelled():
+			return
+		case v := <-sub.Out():
+			ev := v.Data().(tmtypes.EventDataNewBlock)
+			blk = ev.Block
+		}
+
+		// Was block proposed by our node.
+		if bytes.Equal(myAddr, blk.ProposerAddress) {
+			metrics.ProposedBlocks.With(labelTendermint).Inc()
+		}
+
+		// Was block voted for by our node. Ignore if there was no previous block.
+		if blk.LastCommit != nil {
+			for _, a := range blk.LastCommit.Signatures {
+				if bytes.Equal(myAddr, a.ValidatorAddress) {
```

Comment:
This is not correct as it also includes missing votes. If `sig.IsAbsent()` or `sig.BlockIDFlag == tmtypes.BlockIDFlagNil` the vote should not be counted.